src/service/viewport/viewport-binding-ios-embed-sd.js

```diff
@@ -0,0 +1,429 @@
+      // browsers. Thus we'll use the `i-amphtml-ios-overscroll` class for
+      // testing.
+      this.scroller_.classList.add('i-amphtml-ios-overscroll');
+      setImportantStyles(this.scroller_, {
+        '-webkit-overflow-scrolling': 'touch',
+      });
+    });
+
+    dev().fine(TAG_, 'initialized ios-embed-wrapper-sd viewport');
+  }
+
+  /** @override */
+  ensureReadyForElements() {
+    this.setup_();
+  }
+
+  /** @private */
+  setup_() {
+    if (this.setupDone_) {
+      return;
+    }
+    this.setupDone_ = true;
+
+    // Embedded scrolling on iOS is rather complicated. IFrames cannot be sized
+    // and be scrollable. Sizing iframe by scrolling height has a big negative
+    // that "fixed" position is essentially impossible. The only option we
+    // found is to reset scrolling on the AMP doc, which wraps the natural BODY
+    // children inside the `overflow:auto` element inside the shadow root of the
+    // body. For reference, here are related
+    // iOS issues (Chrome issues are also listed for reference):
+    // - https://code.google.com/p/chromium/issues/detail?id=2891
+    // - https://code.google.com/p/chromium/issues/detail?id=157855
+    // - https://bugs.webkit.org/show_bug.cgi?id=106133
+    // - https://bugs.webkit.org/show_bug.cgi?id=149264
+    const doc = this.win.document;
+    const body = dev().assertElement(doc.body, 'body is not available');
+    const shadowRoot = body.attachShadow({mode: 'open'});
+
+    // Main slot will absorb all undistributed children.
+    const mainSlot = doc.createElement('slot');
+    this.wrapper_.appendChild(mainSlot);
+    shadowRoot.appendChild(this.scroller_);
+
+    // Update body styles and monitor for further changes.
+    this.updateBodyStyles_();
+    if (this.win.MutationObserver) {
+      const bodyObserver = new this.win.MutationObserver(
+          this.updateBodyStyles_.bind(this));
+      bodyObserver.observe(body, {attributes: true});
+    }
+
+    // Make sure the scroll position is adjusted correctly.
+    this.onScrolled_();
+  }
+
+  /** @private */
+  updateBodyStyles_() {
+    const {body} = this.win.document;
+    if (!body) {
+      return;
+    }
+
+    // Many `<body>` styles are inherited to ensure that layout is preserved.
+    // The most important: `display` and related styles.
+    const inheritStyles = {};
+    this.vsync_.run({
+      measure: () => {
+        const bodyStyles = computedStyle(this.win, body);
+        INHERIT_STYLES.forEach(style => {
+          inheritStyles[style] = bodyStyles[style];
+        });
+      },
+      mutate: () => {
+        setImportantStyles(this.wrapper_, inheritStyles);
+      },
+    });
+  }
+
+  /** @private */
+  onResized_() {
+    this.resizeObservable_.fire();
+    this.updateBodyStyles_();
+  }
+
+  /** @override */
+  connect() {
+    this.win.addEventListener('resize', this.boundResizeEventListener_);
+    this.scroller_.addEventListener('scroll', this.boundScrollEventListener_);
```

Comment:
Connect and disconnect the mutation observer, too?

Reply:
Not necessary - the body in the ampdoc structure is never shared between several docs in a single window.